### ISSUE TITLE
When used with objects, the return type unifies all value types

### ISSUE THREAD

**Issue by novemberborn** [sindresorhus/p-props#16]
```ts
const foo = await pProps({ bar: 'baz' as const, thud: 'qux' as const })
```

`foo` is:

```ts
type Foo = {
  bar: 'baz' | 'qux',
  thud: 'baz' | 'qux'
}
```

I would have expected it to be:

```ts
type Foo = {
  bar: 'baz',
  thud: 'qux`
}
```

**Comment by sindresorhus** [comment]
If anyone wants to work on this, see the initial attempt and feedback in https://github.com/sindresorhus/p-props/pull/17.

**Comment by ChristianBoehlke** [comment]
We have a lot of type-casts in our codebase as well because of this issue and it looks like #17 would have worked well for us and the problem described here. Could you maybe explain a bit more what else you're looking for to accept a PR?

I understand that there is the advanced case with a mapper function that stayed unchanged in #17, but it did improve the "basic" use case considerably! I would also argue that when a mapper function is used, the types are probably okay like they are now because they would most likely be the same. On the other hand, if the object contains the promises directly the chances are pretty high (at least for us) that the types are different and it would be very beneficial if the return type would narrow them down correctly.

**Comment by karlhorky** [comment]
@ChristianBoehlke are you interested in doing a PR for this? Would be great to get this improvement!

**Comment by karlhorky** [comment]
I was looking for alternatives, and [found a few on npmtrends](https://npmtrends.com/@0xbutterfield/p-props-vs-p-props-vs-promise-all-props-vs-promise-props-vs-promise-props-recursive), but none that are well maintained, popular and written in TypeScript:

- [`promise-props`](https://www.npmjs.com/package/promise-props) by @ide of @expo ([repo](https://github.com/expo/promise-props))
- [`@0xbutterfield/p-props`](https://www.npmjs.com/package/@0xbutterfield/p-props) by @0xButterfield ([repo](https://github.com/0xButterfield/p-props)) - fork that removes the mapper + adjusts types

**Comment by Siilwyn** [comment]
Thanks for the ping @karlhorky, and even more for the types code! I've published https://www.npmjs.com/package/promise-all-props 3.0.0 with types.

**Comment by karlhorky** [comment]
@sindresorhus Thanks for https://github.com/sindresorhus/p-props/commit/7eb631f49e563c5e472d7102ad0ab4a1d0b0b67d 🙌 

This was released as part of [`p-props@6.0.0`](https://github.com/sindresorhus/p-props/releases/tag/v6.0.0). We just switched back to this version and can confirm that it's working 👍
